src/main/java/bio/terra/service/filedata/DrsService.java

```diff
@@ -452,4 +448,46 @@ public static String getLastNameFromPath(String path) {
     String[] pathParts = StringUtils.split(path, '/');
     return pathParts[pathParts.length - 1];
   }
+
+  private void verifyAuthorization(String snapshotId, AuthenticatedUserRequest authUser) {
+    samAuthorizations.computeIfAbsent(
+        snapshotId,
+        id -> {
+          samService.verifyAuthorization(
+              authUser, IamResourceType.DATASNAPSHOT, id, IamAction.READ_DATA);
+          return id;
+        });
+  }
+
+  private SnapshotProject getSnapshotProject(UUID snapshotId) {
+    return snapshotProjects.computeIfAbsent(
+        snapshotId, snapshotService::retrieveAvailableSnapshotProject);
+  }
+
+  private SnapshotCacheResult getSnapshot(UUID snapshotId) {
+    return snapshotCache.computeIfAbsent(
+        snapshotId, id -> new SnapshotCacheResult(snapshotService.retrieve(id)));
+  }
+
+  private static class SnapshotCacheResult {
+    private final Boolean isSelfHosted;
+    private final BillingProfileModel billingProfileModel;
+    private String googleProjectId;
+
+    public SnapshotCacheResult(Snapshot snapshot) {
+      this.isSelfHosted = snapshot.isSelfHosted();
+      this.billingProfileModel =
+          snapshot
+              .getFirstSnapshotSource()
+              .getDataset()
+              .getDatasetSummary()
+              .getDefaultBillingProfile();
+
+      CloudPlatformWrapper platform =
+          CloudPlatformWrapper.of(billingProfileModel.getCloudPlatform());
+      if (platform.isGcp()) {
+        this.googleProjectId = snapshot.getProjectResource().getGoogleProjectId();
+      }
```

Comment:
This seems more complicated than writing
```suggestion
      var projectResource = snapshot.getProjectResource();
      if (projectResource != null) {
        this.googleProjectId = projectResource.getGoogleProjectId();
      }
```

Reply:
On second thought, it's nice to have a marker to search for when looking for cloud platform specific steps.